File: Person4_AIBiometricService/clients/node/pnsmAiClient.js
/**
 * PNSM AI service client for Person 3's Node.js backend.
 *
 * Copy this file into your project. Zero dependencies -- Node 18+ built-ins only.
 * The signing logic here is the reference implementation: if you sign requests
 * a different way, they will be rejected with 401.
 *
 *   const { PnsmAiClient } = require('./pnsmAiClient');
 *
 *   const ai = new PnsmAiClient({
 *     baseUrl: process.env.PNSM_AI_URL,       // the AI service's private DNS name
 *     hmacSecret: process.env.PNSM_HMAC_SECRET // base64, from Person 4
 *   });
 *
 * Owner: Person 4. Questions about the contract go to docs/API.md.
 */

'use strict';

const crypto = require('node:crypto');

/** Thrown for any non-2xx response. Carries the service's reason code. */
class PnsmAiError extends Error {
  constructor(status, body) {
    const error = (body && body.error) || {};
    super(error.message || `PNSM AI service returned ${status}`);
    this.name = 'PnsmAiError';
    this.status = status;
    /** A member of the reason-code table in docs/API.md. Branch on this. */
    this.code = error.code || 'INTERNAL';
    /** True when retrying the identical request could succeed. */
    this.retryable = Boolean(error.retryable);
    this.requestId = error.request_id || null;
    this.context = error.context || null;
  }
}

class PnsmAiClient {
  /**
   * @param {object} options
   * @param {string} options.baseUrl      Service base URL, no trailing slash.
   * @param {string} options.hmacSecret   Base64 32-byte shared secret.
   * @param {string} [options.adminToken] Only needed for /metrics and /v1/admin/*.
   * @param {number} [options.timeoutMs]  Per-request timeout. Default 10000.
   */
  constructor({ baseUrl, hmacSecret, adminToken = '', timeoutMs = 10_000 }) {
    if (!baseUrl) throw new Error('baseUrl is required');
    if (!hmacSecret) throw new Error('hmacSecret is required');
    this.baseUrl = baseUrl.replace(/\/+$/, '');
    this.secret = Buffer.from(hmacSecret, 'base64');
    if (this.secret.length !== 32) {
      throw new Error(`hmacSecret must decode to 32 bytes, got ${this.secret.length}`);
    }
    this.adminToken = adminToken;
    this.timeoutMs = timeoutMs;
  }

  /**
   * Build the two auth headers.
   *
   * The signed payload is `${timestamp}.${rawBody}` -- the timestamp is inside
   * the signature so a captured signature cannot be replayed with a fresh time.
   * Sign the exact bytes you send: re-serialising the object changes them.
   */
  _headers(rawBody, { admin = false } = {}) {
    const timestamp = String(Math.floor(Date.now() / 1000));
    const signature = crypto
      .createHmac('sha256', this.secret)
      .update(`${timestamp}.`)
      .update(rawBody)
      .digest('hex');
    const headers = {
      'Content-Type': 'application/json',
      'X-PNSM-Timestamp': timestamp,
      'X-PNSM-Signature': `v1=${signature}`,
    };
    if (admin && this.adminToken) headers['X-PNSM-Admin'] = this.adminToken;
    return headers;
  }

  async _post(path, payload, options = {}) {
    const rawBody = Buffer.from(JSON.stringify(payload), 'utf8');
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let response;
    try {
      response = await fetch(this.baseUrl + path, {
        method: 'POST',
        headers: this._headers(rawBody, options),
        body: rawBody,
        signal: controller.signal,
      });
    } catch (cause) {
      // A network failure is retryable; surface it in the same shape as
      // everything else so callers need only one error path.
      throw new PnsmAiError(0, {
        error: { code: 'UPSTREAM_STORAGE_ERROR', message: String(cause), retryable: true },
      });
    } finally {
      clearTimeout(timer);
    }

    const body = await response.json().catch(() => ({}));
    if (!response.ok) throw new PnsmAiError(response.status, body);
    return body;
  }

  async _get(path, options = {}) {
    const response = await fetch(this.baseUrl + path, {
      method: 'GET',
      headers: this._headers(Buffer.alloc(0), options),
    });
    const body = await response.json().catch(() => ({}));
    if (!response.ok) throw new PnsmAiError(response.status, body);
    return body;
  }

  /** Liveness. Cheap; safe to poll. Needs no signature but sending one is fine. */
  async health() {
    const response = await fetch(this.baseUrl + '/health');
    return response.json();
  }

  /** Readiness, including model, calibration and warmup state. */
  async ready() {
    const response = await fetch(this.baseUrl + '/ready');
    return response.json();
  }

  /**
   * Onboarding (FR-01). Store `result.envelope` verbatim in the collection
   * named by `result.storage_target` -- FaceEmbeddings, NOT the Users document.
   * Do not reformat it and do not try to read it: the key never leaves the AI
   * service, and under the KMS provider the data key is unwrapped only there.
   *
   * @param {string} userRef    Employee _id as a string.
   * @param {object} image      { kind: 's3_key' | 'base64', value: string }
   * @param {string} requestId  A 26-character ULID.
   */
  async embed(userRef, image, requestId) {
    return this._post('/v1/embed', { user_ref: userRef, image, request_id: requestId });
  }

  /**
   * Check-in (FR-05, FR-07).
   *
   * Returns a decision, not an exception, whenever a comparison completed:
   * `approved`, `rejected`, or -- only under the three-band policy -- `flagged`.
   * Everything else throws a PnsmAiError; branch on `err.code`.
   *
   * Push the WebSocket alert off `result.hr_alert`, NOT off the decision
   * string. It is true for anything short of a clean approval under either
   * band policy, so the alerting survives a policy change.
   *
   * `confidence` is the calibrated similarity score as a percentage, not
   * `cosine * 100`; `raw_cosine` is the audit trail and must never be shown to
   * an employee.
   */
  async verify({ userRef, image, envelope, requestId, capturedAt, device }) {
    return this._post('/v1/verify', {
      user_ref: userRef,
      image,
      envelope,
      request_id: requestId,
      captured_at: capturedAt instanceof Date ? capturedAt.toISOString() : capturedAt,
      device: {
        platform: device?.platform ?? 'unknown',
        os_version: device?.osVersion ?? '',
        app_version: device?.appVersion ?? '',
        is_mock_location: Boolean(device?.isMockLocation),
        is_emulator: Boolean(device?.isEmulator),
        is_rooted: Boolean(device?.isRooted),
      },
    });
  }

  /** Hash a new PIN (FR-06). Store the whole record on the employee document. */
  async hashPin(userRef, pin) {
    return this._post('/v1/security/pin/hash', { user_ref: userRef, pin });
  }

  /**
   * Check a PIN. A wrong PIN is a normal 200 with `match: false`; once the
   * attempt budget is spent the call throws RATE_LIMITED and no comparison is
   * performed at all.
   */
  async verifyPin(userRef, pin, pinHash) {
    return this._post('/v1/security/pin/verify', {
      user_ref: userRef,
      pin,
      pin_hash: pinHash,
    });
  }

  /**
   * Mint an upload URL for the mobile client or the dashboard.
   * `contentLength` must be the exact compressed byte count -- it is signed
   * into the URL, so the bucket rejects a mismatch.
   */
  async presignPut({ userRef, purpose, objectId, contentType, contentLength }) {
    return this._post('/v1/storage/presign-put', {
      user_ref: userRef,
      purpose,
      object_id: objectId,
      content_type: contentType,
      content_length: contentLength,
    });
  }

  /** Short-lived read URL for the HR audit view. */
  async presignGet(objectKey) {
    return this._post('/v1/storage/presign-get', { object_key: objectKey });
  }

  /** Reload calibration.json without a restart. Requires the admin token. */
  async recalibrate() {
    return this._post('/v1/admin/recalibrate', {}, { admin: true });
  }

  /** Counters, latency summary and resident memory. Requires the admin token. */
  async metrics() {
    return this._get('/metrics', { admin: true });
  }
}

/**
 * Crockford base32 ULID. Use one per check-in: it is the idempotency key and
 * the replay guard, so it must be fresh for every genuine attempt and reused
 * only when retrying the identical request.
 */
function newUlid() {
  const ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
  let time = Date.now();
  let timePart = '';
  for (let i = 0; i < 10; i += 1) {
    timePart = ALPHABET[time % 32] + timePart;
    time = Math.floor(time / 32);
  }
  const random = crypto.randomBytes(16);
  let randomPart = '';
  for (let i = 0; i < 16; i += 1) randomPart += ALPHABET[random[i] % 32];
  return timePart + randomPart;
}

module.exports = { PnsmAiClient, PnsmAiError, newUlid };

File: M/ai-service/clients/python/pnsm_ai_client.py
"""PNSM AI service client for Person 3's Python/FastAPI backend.

Copy this file into your project.  Depends only on ``httpx``.  The signing
logic here is the reference implementation: sign requests any other way and
they come back 401.

    from pnsm_ai_client import PnsmAiClient, PnsmAiError, new_ulid

    ai = PnsmAiClient(
        base_url=os.environ["PNSM_AI_URL"],
        hmac_secret=os.environ["PNSM_HMAC_SECRET"],
    )

    result = await ai.verify(
        user_ref=str(user["_id"]),
        image={"kind": "s3_key", "value": object_key},
        envelope=face_doc["envelope"],   # from the FaceEmbeddings collection
        request_id=new_ulid(),
        captured_at=payload.captured_at,
        device=payload.device,
    )
    if result["decision"] == "approved":
        ...

Owner: Person 4. The contract lives in docs/API.md.
"""

from __future__ import annotations

import base64
import datetime as _dt
import hashlib
import hmac
import json
import os
import secrets
import time
from collections.abc import Mapping
from typing import Any, Literal

import httpx

ULID_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"


class PnsmAiError(Exception):
    """Any non-2xx response. Carries the service's reason code."""

    def __init__(self, status: int, body: Mapping[str, Any] | None = None) -> None:
        error = dict((body or {}).get("error") or {})
        super().__init__(error.get("message") or f"PNSM AI service returned {status}")
        self.status = status
        #: A member of the reason-code table in docs/API.md. Branch on this.
        self.code: str = error.get("code", "INTERNAL")
        #: True when retrying the identical request could succeed.
        self.retryable: bool = bool(error.get("retryable"))
        self.request_id: str | None = error.get("request_id")
        self.context: dict[str, Any] | None = error.get("context")


def new_ulid() -> str:
    """A Crockford base32 ULID.

    Use a fresh one per genuine check-in: it is both the idempotency key and
    the replay guard. Reuse it only when retrying the *identical* request.
    """
    milliseconds = int(time.time() * 1000)
    time_part = ""
    for _ in range(10):
        time_part = ULID_ALPHABET[milliseconds % 32] + time_part
        milliseconds //= 32
    random_part = "".join(secrets.choice(ULID_ALPHABET) for _ in range(16))
    return time_part + random_part


class PnsmAiClient:
    """Async client. Construct once and reuse -- it holds a connection pool."""

    def __init__(
        self,
        *,
        base_url: str,
        hmac_secret: str,
        admin_token: str = "",
        timeout_s: float = 10.0,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._secret = base64.b64decode(hmac_secret)
        if len(self._secret) != 32:
            raise ValueError(f"hmac_secret must decode to 32 bytes, got {len(self._secret)}")
        self._admin_token = admin_token
        self._client = httpx.AsyncClient(base_url=self._base_url, timeout=timeout_s)

    async def aclose(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> PnsmAiClient:
        return self

    async def __aexit__(self, *_: object) -> None:
        await self.aclose()

    # ---------------------------------------------------------------- signing
    def _headers(self, body: bytes, *, admin: bool = False) -> dict[str, str]:
        """The signed payload is ``f"{timestamp}.{raw_body}"``.

        The timestamp is inside the signature so a captured signature cannot be
        replayed with a fresh time. Sign the exact bytes you send -- re-encoding
        the JSON changes them and invalidates the signature.
        """
        timestamp = str(int(time.time()))
        digest = hmac.new(
            self._secret, timestamp.encode("ascii") + b"." + body, hashlib.sha256
        ).hexdigest()
        headers = {
            "Content-Type": "application/json",
            "X-PNSM-Timestamp": timestamp,
            "X-PNSM-Signature": f"v1={digest}",
        }
        if admin and self._admin_token:
            headers["X-PNSM-Admin"] = self._admin_token
        return headers

    async def _post(self, path: str, payload: dict[str, Any], *, admin: bool = False) -> dict[str, Any]:
        body = json.dumps(payload).encode("utf-8")
        try:
            response = await self._client.post(path, content=body, headers=self._headers(body, admin=admin))
        except httpx.HTTPError as exc:
            # Surface transport failures in the same shape as everything else
            # so callers need exactly one error path.
            raise PnsmAiError(
                0,
                {"error": {"code": "UPSTREAM_STORAGE_ERROR", "message": str(exc), "retryable": True}},
            ) from exc
        return self._unwrap(response)

    async def _get(self, path: str, *, admin: bool = False) -> dict[str, Any]:
        response = await self._client.get(path, headers=self._headers(b"", admin=admin))
        return self._unwrap(response)

    @staticmethod
    def _unwrap(response: httpx.Response) -> dict[str, Any]:
        try:
            body = response.json()
        except ValueError:
            body = {}
        if response.status_code >= 400:
            raise PnsmAiError(response.status_code, body)
        return body

    # -------------------------------------------------------------------- ops
    async def health(self) -> dict[str, Any]:
        return self._unwrap(await self._client.get("/health"))

    async def ready(self) -> dict[str, Any]:
        return self._unwrap(await self._client.get("/ready"))

    # ------------------------------------------------------------- biometrics
    async def embed(
        self, *, user_ref: str, image: Mapping[str, str], request_id: str | None = None
    ) -> dict[str, Any]:
        """Onboarding (FR-01).

        Store ``result["envelope"]`` verbatim on the employee document. Do not
        reformat it and do not try to read it: the key never leaves the AI
        service, which is what keeps a bug in your code from leaking biometrics.
        """
        return await self._post(
            "/v1/embed",
            {"user_ref": user_ref, "image": dict(image), "request_id": request_id or new_ulid()},
        )

    async def verify(
        self,
        *,
        user_ref: str,
        image: Mapping[str, str],
        envelope: Mapping[str, Any],
        request_id: str,
        captured_at: _dt.datetime | str,
        device: Mapping[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Check-in (FR-05, FR-07).

        Returns a decision whenever a comparison completed: ``approved``,
        ``rejected``, or -- only when the service runs the three-band policy --
        ``flagged``. Anything else raises :class:`PnsmAiError`; branch on
        ``err.code``.

        **Push the HR alert off ``result["hr_alert"]``, not off the decision
        string.** It is true for anything short of a clean approval under either
        band policy, so the alerting keeps working if the policy changes.

        ``confidence`` is the calibrated similarity score as a percentage --
        not ``cosine * 100``. ``raw_cosine`` is the audit trail and must never
        be shown to an employee.

        Pass the device flags through unchanged. They are the FR-05 control,
        and a field you forget to forward is a check the service cannot make.
        """
        stamp = captured_at.isoformat() if isinstance(captured_at, _dt.datetime) else captured_at
        supplied = dict(device or {})
        return await self._post(
            "/v1/verify",
            {
                "user_ref": user_ref,
                "image": dict(image),
                "envelope": dict(envelope),
                "request_id": request_id,
                "captured_at": stamp,
                "device": {
                    "platform": supplied.get("platform", "unknown"),
                    "os_version": supplied.get("os_version", ""),
                    "app_version": supplied.get("app_version", ""),
                    "is_mock_location": bool(supplied.get("is_mock_location", False)),
                    "is_emulator": bool(supplied.get("is_emulator", False)),
                    "is_rooted": bool(supplied.get("is_rooted", False)),
                },
            },
        )

    # ----------------------------------------------------------------- security
    async def hash_pin(self, *, user_ref: str, pin: str) -> dict[str, Any]:
        return await self._post("/v1/security/pin/hash", {"user_ref": user_ref, "pin": pin})

    async def verify_pin(self, *, user_ref: str, pin: str, pin_hash: str) -> dict[str, Any]:
        return await self._post(
            "/v1/security/pin/verify",
            {"user_ref": user_ref, "pin": pin, "pin_hash": pin_hash},
        )

    # ------------------------------------------------------------------ storage
    async def presign_put(
        self,
        *,
        user_ref: str,
        purpose: Literal["reference", "checkin"],
        object_id: str,
        content_type: Literal["image/webp", "image/jpeg"],
        content_length: int,
    ) -> dict[str, Any]:
        """``content_length`` must be the exact compressed byte count.

        It is signed into the URL, so the bucket itself rejects a mismatch --
        the size ceiling is enforced by infrastructure, not by client goodwill.
        """
        return await self._post(
            "/v1/storage/presign-put",
            {
                "user_ref": user_ref,
                "purpose": purpose,
                "object_id": object_id,
                "content_type": content_type,
                "content_length": content_length,
            },
        )

    async def presign_get(self, *, object_key: str) -> dict[str, Any]:
        return await self._post("/v1/storage/presign-get", {"object_key": object_key})

    # -------------------------------------------------------------------- admin
    async def recalibrate(self) -> dict[str, Any]:
        return await self._post("/v1/admin/recalibrate", {}, admin=True)

    async def metrics(self) -> dict[str, Any]:
        return await self._get("/metrics", admin=True)


def from_environment() -> PnsmAiClient:
    """Build a client from ``PNSM_AI_URL`` / ``PNSM_HMAC_SECRET``."""
    return PnsmAiClient(
        base_url=os.environ["PNSM_AI_URL"],
        hmac_secret=os.environ["PNSM_HMAC_SECRET"],
        admin_token=os.environ.get("PNSM_ADMIN_TOKEN", ""),
    )
